File: C++/permutation.cpp
/**
 * @File    : permutation.cpp
 * @Brief   : 字符串的排列
 * @Link    : https://leetcode-cn.com/problems/zi-fu-chuan-de-pai-lie-lcof/
 * @Author  : Wei Li
 * @Date    : 2021-09-29
*/

#include <iostream>
#include <string>
#include <vector>
#include <algorithm>

class Solution
{
public:
    bool nextPermutation(std::string &s)
    {
        int i = s.size() - 2;
        while (i >= 0 && s[i] >= s[i + 1])
        {
            i--;
        }
        if (i < 0)
        {
            return false;
        }
        int j = s.size() - 1;
        while (j >= 0 && s[i] >= s[j])
        {
            j--;
        }
        std::swap(s[i], s[j]);
        reverse(s.begin() + i + 1, s.end());
        return true;
    }

    std::vector<std::string> permutation(std::string s)
    {
        std::vector<std::string> ret;
        sort(s.begin(), s.end());
        do
        {
            ret.push_back(s);
        } while (nextPermutation(s));
        return ret;
    }
};

template <typename T>
void DisplayElement(const T &container)
{
    for (auto element : container)
    {
        std::cout << element << " ";
    }
    std::cout << std::endl;
}

// ----------------------------
int main(int argc, char **argv)
{
    std::string str = "abc";

    Solution solution;
    auto str_permutation = solution.permutation(str);
    std::cout << "The solution of this problem is " << std::endl;
    DisplayElement(str_permutation);

    return 0;
}

File: Python/permutation.py
#!/usr/bin/env python3
# encoding: utf-8

"""
@Filename: permutation.py
@Function: 字符串的排列
@Link: https://leetcode-cn.com/problems/zi-fu-chuan-de-pai-lie-lcof/
@Python Version: 3.8
@Author: Wei Li
@Date：2021-09-29
"""


from typing import List

class Solution:
    def permutation(self, s: str) -> List[str]:
        c, res = list(s), []
        def dfs(x):
            if x == len(c) - 1:
                res.append(''.join(c))   # 添加排列方案
                return
            dic = set()
            for i in range(x, len(c)):
                if c[i] in dic: continue # 重复，因此剪枝
                dic.add(c[i])
                c[i], c[x] = c[x], c[i]  # 交换，将 c[i] 固定在第 x 位
                dfs(x + 1)               # 开启固定第 x + 1 位字符
                c[i], c[x] = c[x], c[i]  # 恢复交换
        dfs(0)
        return res


# -------------------------
if __name__ == "__main__":
    str = "abc"

    solution = Solution()
    str_permutation = solution.permutation(str)
    print(f"The solution of this problem is \n{str_permutation}")
